src/lib/integration_test_lib/malleable_error.ml

```diff
@@ -0,0 +1,345 @@
+open Core_kernel
+open Async_kernel
+
+(* module Test_error = Test_error *)
+
+(** The is a monad which is conceptually similar to `Deferred.Or_error.t`,
+ *  except that there are 2 types of errors which can be returned at each bind
+ *  point in a computation: soft errors, and hard errors. Soft errors do not
+ *  effect the control flow of the monad, and are instead accumulated for later
+ *  extraction. Hard errors effect the control flow of the monad in the same
+ *  way an `Error` constructor for `Or_error.t` would.
+ It remains similar to Deferred.Or_error.t in that it is a specialization of Deferred.Result.t
+ *)
+
+module Hard_fail = struct
+  type t =
+    { hard_error: Test_error.internal_error
+    ; soft_errors: Test_error.internal_error list }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module Accumulator = struct
+  type 'a t =
+    {computation_result: 'a; soft_errors: Test_error.internal_error list}
+  [@@deriving eq, sexp_of, compare]
+end
+
+type 'a t = ('a Accumulator.t, Hard_fail.t) Deferred.Result.t
+
+module T = Monad.Make (struct
+  type nonrec 'a t = 'a t
+
+  let return a =
+    Deferred.return (Ok {Accumulator.computation_result= a; soft_errors= []})
+
+  (* 
+
+('a, 'b) result Deferred.t ->
+f:('c -> ('d, 'e) result Deferred.t) -> 
+('f, 'b) result Deferred.t 
+
+*)
+  let bind res ~f =
+    match%bind res with
+    | Ok {Accumulator.computation_result= prev_result; soft_errors= err_list}
+      -> (
+        let execProg = f prev_result in
+        match%bind execProg with
+        | Ok {Accumulator.computation_result= comp; soft_errors= next_list} ->
+            Deferred.return
+              (Ok
+                 { Accumulator.computation_result= comp
+                 ; Accumulator.soft_errors= List.append err_list next_list })
+        | Error
+            {Hard_fail.hard_error= hard_err; Hard_fail.soft_errors= next_list}
+          ->
+            Deferred.return
+              (Error
+                 { Hard_fail.hard_error= hard_err
+                 ; soft_errors= List.append err_list next_list }) )
+    | Error _ as error ->
+        Deferred.return error
+
+  let map = `Define_using_bind
+end)
+
+include T
+
+let return_without_deferred a =
```

Comment:
I think these `*_without_deferred` aren't really necessary and just cause a little bit of confusion (and an abstraction leak). Looking at all the places this is used, it seems like those can all be rewritten without this function, so I would recommend doing that. For example:

```ocaml
let _ = Deferred.map (Ivar.read init) ~f:Malleable_error.return_without_deferred
(* can be rewritten as *)
let _ = Deferred.bind (Ivar.read init) ~f:Malleable_error.return
```